falco: Add base plugin

knrt10 opened the pull request:
Add plugin for Falco.

### TODO

- [ ] Find a way and fix to get all rules configured by Falco

### How to test

```bash
# clone Headlamp
git clone git@github.com:kubernetes-sigs/headlamp.git

# Run headlamp backend
cd backend && go run ./cmd -dev -enable-dynamic-clusters -proxy-urls=https://artifacthub.io/\*

# Run frontend
cd frontend && npm i && npm start

# Run app
cd app && npm run dev-only-app

# Have a k8s cluster up and running. See the readme of this PR

# Fetch this PR and go to falco directory
cd falco && npm i && npm start

```

### Screenshots

<img width="1736" alt="Screenshot 2025-05-13 at 1 28 14 PM" src="https://github.com/user-attachments/assets/9a78b6f0-4aad-4910-85a0-03388c45bb06" />
<img width="1740" alt="Screenshot 2025-05-13 at 1 28 27 PM" src="https://github.com/user-attachments/assets/2369aaf7-c882-42fc-a5bb-60bba3e49cc1" />
<img width="1390" alt="Screenshot 2025-05-13 at 1 28 39 PM" src="https://github.com/user-attachments/assets/6eb28c18-f8ea-46b9-b8fa-31f4d947eb92" />
<img width="1496" alt="Screenshot 2025-05-13 at 1 28 52 PM" src="https://github.com/user-attachments/assets/0b9738dd-e058-4871-82a3-ac147ece0afc" />
<img width="1456" alt="Screenshot 2025-05-13 at 1 29 04 PM" src="https://github.com/user-attachments/assets/e1fc443c-1f1b-4d99-b7ee-a979dd7927c5" />

Commit:
falco: Add base plugin

This adds events, rule and settings page to the plugin. User can select
the backend for the events. Default backend is file/volumes. They can
easily configure redis as well and use that as backend.

Signed-off-by: Kautilya Tripathi <ktripathi@microsoft.com>

gambtho commented on falco/redis/redis-rest-proxy.yaml:
```diff
@@ -0,0 +1,90 @@
+      name: http
+  selector:
+    app: redis-rest-proxy
+
```
consider restricting to clusterIP or adding a network policy? currently i think anyone can read all events/inject fake events

illume reviewed (changes requested):
<!-- prch:feedback -->
Thanks for these changes.

<!-- prch:git-conflicts -->
It looks like this PR has git conflicts. Can you please fix them?

illume commented:
@copilot+claude-opus-4.6 address open review comments

Copilot committed:
Resolve merge conflict in README.md

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: return fresh copy of defaultSettings in loadSettings

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: pass pendingSettings.redisUrl to RedisConnectionTester

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: reuse shared utilities from storageUtils and constants

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: cap file-stream events array to prevent unbounded growth

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: replace deprecated onKeyPress with onKeyDown

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: include default namespace in getNamespace extraction

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: scope pod listing to FALCO_NAMESPACE

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: preserve Unicode in cleanYamlText, only strip control chars

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: handle both namespace and namespaces kind in K8sSourceCell

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: use @property instead of @param in FalcoEvent JSDoc

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: pin Python dependencies in redis-rest-proxy

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: add unit tests for utility functions

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: add Storybook stories for components

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: fix lint issues in tests and stories

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: add tests and stories

Agent-Logs-Url: https://github.com/headlamp-k8s/plugins/sessions/f37e2a66-7f0d-4912-84fe-550e61db55c0

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: add .eslintcache to .gitignore

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: cancel file stream exec on cleanup

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: harden redis rest proxy manifest

Co-authored-by: illume <9541+illume@users.noreply.github.com>

illume commented:
@copilot+claude-opus-4.7 address open review comments

Copilot committed:
falco: remove debug console.log statements from rules loaders

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: allow API server ingress to redis-rest-proxy NetworkPolicy

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: validate and normalize Redis URL, reuse FALCO_NAMESPACE in tester

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: validate redisUrl, decode ArrayBufferView correctly, detect Falco container name

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: map kinds to Headlamp routeNames explicitly in K8sSourceCell

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: preserve URL hash fragment in normalizeRedisUrl

Agent-Logs-Url: https://github.com/headlamp-k8s/plugins/sessions/d4ccc1bc-0993-4000-8406-a019a52c3952

Co-authored-by: illume <9541+illume@users.noreply.github.com>

illume commented:
@copilot+claude-opus-4.7 Address open review comments

Copilot committed:
falco: bump @kinvolk/headlamp-plugin to ^0.13.1 and add i18n script

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: clear save-status setTimeout on unmount

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: handle ns/ms/s epoch and short namespace names in falcoEventUtils

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: fix non-root pip install and cap falco:events with LTRIM

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: parallelize fetchAllRules with Promise.all

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: remove duplicate path from logo SVG

Co-authored-by: illume <9541+illume@users.noreply.github.com>

Copilot committed:
falco: wire useFilterFunc into FalcoRules header search

Co-authored-by: illume <9541+illume@users.noreply.github.com>
